Support for IdP initiated proxy SLO

m0ark opened the pull request:
# Problem
SimpleSAMLphp was not designed to support protocol proxy setups. Fortunately it supports a lot of those use cases.
One particular feature that has always been missing in proxy scenarios is the capability to properly propagate logout (SLO) requests coming from an upstream identity provider. This is because of SimpleSAMLphp IdPs using its SPs as authentication sources. In case of an IdP initiated logout the proxy SP is notified about the logout request and terminates its session properly without knowing about components it has been used by and thus cannot propagate the logout request to. 
The result of this behaviour is the users IdP receiving a logout response telling the SLO has properly finished but the proxy IdP(s) still having an active session.

This issue has been discussed in #65 and @relaxnow made a great job implementing a first draft to tackle it (https://github.com/relaxnow/simplesamlphp/commit/35cffcca61c02cb2a01e152e801ad3515c3e5803).
Nevertheless this implementation was way too tightly coupled to SAML components and lacked a proper differentiation between multiple identity providers.

# Solution
This patch is based on the work of @relaxnow and adds a small framework for adding IdP initiated proxy SLO capabilities. It also implements the framework for SAML SPs and SAML/ADFS IdPs to support proxy SLO flows using frontend channels.
It should be possible to add proxy SLO capabilities to other IdP/SP protocol implementations (e.g. OIDC, CAS) using this framework without much of a hassle.

To not break the current behaviour proxy SLO has to be activated by setting 
`‘proxyLogoutEnable’ => true`
in each SP configuration.

## Components of the framework
•	`registerAsConsumerLogoutCallback`
To be called by an authentication source consumer (e.g. IdP). The callback should handle everything that is needed to logout the consumer from its own perspective. In case of an IdP this could be logging out of all associated SPs; in case of a PHP application this could terminate the application local session.
•	`hasAsConsumerLogoutCallbacks`
Check if authentication source consumer logout callbacks have been registered for this source.
•	`callAsConsumerLogoutCallbacks`
To be called by a regular authentication source logout routine (e.g. SAML SP SLO handler) to trigger execution of the registered logout callbacks.
•	`handleAsConsumerLogoutCallbacks`
Only used by `callAsConsumerLogoutCallbacks` and the returnTo page that is called in case of the callback not returning immediately (e.g. redirecting, etc).

m0ark committed:
add callback infrastructure to enable idp initiated / proxy mode logout

m0ark committed:
fix missing LR destination in saml2-logout-resume.php

m0ark committed:
add flag for idp logout to skip logging out of authentication source.
this is used in proxy mode logout scenario to prevent logging out the authentication source twice.

m0ark committed:
remove code formerly used to work around the missing proxy SLO.

m0ark committed:
move registering of asConsumerLogoutCallback to postAuth

m0ark committed:
add missing php doc

thijskh commented on lib/SimpleSAML/IdP.php:
```diff
@@ -90,13 +90,15 @@ private function __construct(string $id)
             }
             $this->config = $metadata->getMetaDataConfig(substr($id, 5), 'adfs-idp-hosted');
 
+            /*
```
This seems misplaced

m0ark replied:
You're right. This is just a left over of a workaround of a bug we discovered.
I will remove it from this PR and raise a new issue to resolve this.

m0ark replied:
The issue has been raised #1564 and the change been reverted.

tvdijen replied:
Either remove the commented code or use it...   Or was this something my rebase raised?

m0ark replied:
This must have been raised by the rebase. I removed it earlier.

thijskh commented on modules/saml/lib/Auth/Source/SP.php:
```diff
@@ -95,13 +102,20 @@ public function __construct($info, $config)
         $this->idp = $this->metadata->getString('idp', null);
         $this->discoURL = $this->metadata->getString('discoURL', null);
         $this->disable_scoping = $this->metadata->getBoolean('disable_scoping', false);
+        $this->proxyLogoutEnabled = $this->metadata->getBoolean('proxyLogoutEnable', false);
```
A new metadata option appears but the docs are not updated to mention it.

m0ark replied:
My bad. Thanks for pointing out.  
I just added the missing option documentation.

thijskh reviewed (commented):
Hi! Thanks for the thorough PR.

I have not reviewed it line by line (yet), did add a few observations. And I notice the base branch is simplesamlphp-1.19. It seems unlikely that we'd merge such a feature there and rather target master.

m0ark commented:
Thanks for your first review.
I implemented this change using the version we used in production. So I just wanted to discuss the probability of having the PR accepted first before putting in more effort and porting this change to master.

thijskh commented on lib/SimpleSAML/Auth/Source.php:
```diff
@@ -218,6 +218,116 @@ public function initLogin($return, $errorURL = null, array $params = [])
+     */
+    public function registerAsConsumerLogoutCallback($callback, $params = []) {
+        assert(is_callable($callback));
+        assert(is_array($params));
```
We can use type hints here and a return type instead of the assertions. This also goes for other methods: use type hints/return types, and remove assertions that only assert the type.

m0ark replied:
Fixed in 7693d74c2d15801603ea39ab6b0a0aaf78122f45

thijskh commented on lib/SimpleSAML/Auth/Source.php:
```diff
@@ -218,6 +218,116 @@ public function initLogin($return, $errorURL = null, array $params = [])
+        $callbacks = $session->getData('\SimpleSAML\Auth\Source.AsConsumerLogoutCallbacks', $this->getAuthId());
+        if ($callbacks === null) {
+            return false;
+        }
```
null is also empty, so this test+return is unneeded.

m0ark replied:
Fixed in f30ec019c07716d048790fd3292be6b4507db34a

thijskh commented on lib/SimpleSAML/Auth/Source.php:
```diff
@@ -218,6 +218,116 @@ public function initLogin($return, $errorURL = null, array $params = [])
+            $returnToUrl = Module::getModuleURL('core/source/as_logout_resume.php', ['id' => $id]);
+
+            foreach ($callbacks as $cb) { // FIXME: use loop variable or get rid of it!
+                $callback = array_shift($callbacks);
```
Yes, please fix this

m0ark replied:
Fixed in 82f1b492009472bdb6b866beada30efada687b4b

thijskh commented on modules/saml/www/sp/saml2-logout-resume.php:
```diff
@@ -0,0 +1,59 @@
+$lr->setInResponseTo($message->getId());
+
+if ($numLoggedOut < count($sessionIndexes)) {
+    \SimpleSAML\Logger::warning('Logged out of ' . $numLoggedOut . ' of ' . count($sessionIndexes) . ' sessions.');
```
Why is this a warning?

m0ark replied:
It's just a copy of the working code of saml2-logout.php.
I think it is not expected to have any active session indexes left of the current SLO request after reaching this point.

thijskh reviewed (commented):
I've checked the code and believe it to be OK. However, I don't have a lot of experience with the scenario in question nor do I have a representative environment. So I'd like to add the caveat that I have not verified it to actually work myself. If you use this in actual production I'd be quite confident to merge it though.

Did you give any consideration as to whether at least some basic tests could be added for this?

m0ark committed:
add missing type hints and remove type assertions

m0ark committed:
remove redundant check

m0ark committed:
optimize code structure of callback loop for better readability

m0ark commented:
Thank you for your thorough review and your helpful annotations.

Do you have any pointers on where to add tests for this code?

For an environment just create a simple SAML proxy and trigger an IdP initiated SLO request.
Without enabling this feature the logout request will not propagate to the proxy IdP component.

![image](https://user-images.githubusercontent.com/33390109/150747267-b5f3a61f-3a51-478a-90fb-998f7e7a83c8.png)

tvdijen commented:
Also closes #1565

tvdijen committed:
Coding style

tvdijen committed:
Coding style

tvdijen commented on lib/SimpleSAML/IdP.php:
```diff
@@ -484,9 +520,11 @@ public function handleLogoutRequest(array &$state, ?string $assocId): void
         $returnTo = Module::getModuleURL('core/idp/resumelogout.php', ['id' => $id]);
 
-        $this->authSource->logout($returnTo);
+        if (!$skipAsLogout) {
```
I'd like to see strict comparison here

tvdijen commented on lib/SimpleSAML/IdP.php:
```diff
@@ -484,9 +520,11 @@ public function handleLogoutRequest(array &$state, ?string $assocId): void
+        }
 
-        if ($assocId !== null) {
+        if ($assocId !== null || $skipAsLogout) {
```
And here

tvdijen commented:
I hope I didn't break anything, but I've rebased the lot against `master`..

thijskh commented:
I think the rebasing was not entirely accurate, at least i've spotted the reintroduction of this code: https://github.com/simplesamlphp/simplesamlphp/pull/1553/files#diff-60baa677d9f59b02cd5b3bbd371583dc05eccd0cb0bb74f2003398031fd84f8dR109-R111

tvdijen committed:
Replace assertions

tvdijen committed:
Fix rebase

tvdijen commented:
Good catch! I've fixed that..  Something else that needs to change is the www-script.. It should be converted to a Controller

tvdijen committed:
Fix rebase: Remove commented code

tvdijen committed:
Bump php version

thijskh committed:
Formatting

thijskh committed:
Reorder some toptics in the index page

thijskh committed:
Improve modules section of docs index a bit

thijskh committed:
update links to external module docs

tvdijen committed:
Add job to trigger docs build

thijskh committed:
Remove obsolete translation instructions

(better none than wrong info)

tvdijen committed:
Remove deprecated string use-case for NameIDPolicy

tvdijen committed:
Perform spell checking

tvdijen committed:
Fix markdown

tvdijen committed:
Fix many spelling mistakes

tvdijen committed:
Fix spelling mistake

tvdijen committed:
Fix for bin/get-translation-strings

tvdijen committed:
Run bin/get-translatable-strings

tvdijen committed:
Fix typo

tvdijen committed:
Merge pull request #1665 from simplesamlphp/feature/trigger-docs-build

Add job to trigger docs build

tvdijen committed:
Fix: build docs on push

tvdijen committed:
Import changelog from 1.19

tvdijen committed:
Branch off changelog

tvdijen committed:
Fix release branches

thijskh committed:
Update url in docs for modern paths and do not silence all curl output

In principle the cron task should only output on error which you
want to know about. Add curl -s to disable curl progressbar but -S
to do show any http errors that occurred.

dependabot[bot] committed:
Bump terser from 5.10.0 to 5.14.2

Bumps [terser](https://github.com/terser/terser) from 5.10.0 to 5.14.2.
- [Release notes](https://github.com/terser/terser/releases)
- [Changelog](https://github.com/terser/terser/blob/master/CHANGELOG.md)
- [Commits](https://github.com/terser/terser/commits)

---
updated-dependencies:
- dependency-name: terser
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

tvdijen committed:
Bypass Symfony check for writable log-dir

Bypass Symfony check for writable log-dir for the case where we configured syslog/errlog/stderr

tvdijen committed:
Fix release-branch

Berdir committed:
Symfony6 (#1675)

Add Symfony 6 support

tvdijen committed:
Fix uninitialized var

Berdir committed:
Also update symfony/dependency-injection (#1677)

Berdir committed:
Translation contracts (#1678)

* Allow symfony/translation-contracts ^3

Co-authored-by: Tim van Dijen <tvdijen@gmail.com>

tvdijen committed:
Bugfix: replace /dev/null with system tmp-dir

tvdijen committed:
Delete simplesamlphp-changelog.md

Remove stray file. Changelog is in the docs-dir

thijskh committed:
Improve debugging message of invalid URLs

thijskh committed:
Extend key rollover docs

tvdijen committed:
Fix markdown

tvdijen committed:
Do not run CI if only md-files were changed

tvdijen committed:
Fix metadata endpoints

thijskh committed:
Slightly improve appearance of expired entity flagging

thijskh committed:
Fix debug info text translation

thijskh committed:
Reduce line length

tvdijen committed:
phpcs: fix multiline statements + indentation

thijskh committed:
Mention new location of admin interface

Closes: #1683

thijskh committed:
Add ETag and check for modifiedness of metadata

thijskh committed:
Need Request now to check last modified status

thijskh committed:
Repair broken (but passing) tests and add more coverage

The protectindexpage test was broken and never changed the
protectindexpage setting, but the logic was also broken so
it always passed and never got past the protected index
page guard in the controller so not a lot of code in the
controller was run.

Also add more tests for the metadata controllers.

thijskh committed:
Fix some bugs in serialize handler

kikz0rsk committed:
Add Slovak translation (#1688)

* Translated base attributes and messages to slovak

* Translated module admin to slovak

* Translated modules core and saml to slovak

* Replace tabs with spaces

Co-authored-by: Christian Danizek <christian@danizek.com>

thijskh committed:
De-uglify login form when checkbox shown

thijskh committed:
Fix output of phpinfo controller

RunnableResponse is not right since phpinfo will output immediately
and returns a bool.

thijskh committed:
Implement phpinfo controller as a StreamedResponse

tvdijen committed:
Shorten StreamedResponse call

A callback doesn't necessarily have to be an anonymous function

tvdijen committed:
Add note on changing endpoints

tvdijen committed:
Retrieve relayState from the proper parameter bag

tvdijen committed:
Update security-checker

thijskh committed:
Google Workspace signout page is not SAML SLO

thijskh committed:
Add more coverage for MetadataStorageHandling.

Also do not restrict coverage calculation to this class since
the metada functions to test the various configured subclasses.

tvdijen committed:
Bump dependencies

tvdijen committed:
Fix tests

tvdijen committed:
Bump dependencies

tvdijen committed:
Fix incorrect method call

thijskh committed:
Keep the ACS url as it was in 1.19. (#1698)

But implement it in a way that still does not require any legacy
www scripts.

tvdijen committed:
Bump dependencies

fengtan committed:
Add support for Redis-Sentinel (#1699)

* Add support for Redis-Sentinel.

* Add tests for redis sentinel store.

* Documentation about Redis-Sentinel and password protection.

* Fix markdown syntax brought up by github actions.

* Add missing config parameter to unit tests.

* Add missing return-type

* Fix indentation

* Fix incorrect configuration-methods

Co-authored-by: Tim van Dijen <tvdijen@gmail.com>

tvdijen committed:
Merge branch 'simplesamlphp-2.0'

tvdijen committed:
Fix

tvdijen committed:
Fix build

m0ark committed:
make sure that same associationGroup is only used if both ADFS and SAML IdPs are enabled (#1565)

tvdijen committed:
Merge branch 'master' into proxy-logout-v1.19.1

tvdijen committed:
WIP: Migrate www-scripts to controllers

tvdijen commented:
I'm not sure how to fix the conflicts in `src/SimpleSAML/IdP.php` because the code has been refactored for Symfony responses
